<h2><center>IBM Employee Attrition ML Project</center></h2>
<h2><center>Exploratory Data Analysis</center></h2>
<h4><center>Author: Akshay Pandurang Paunikar</center></h4>

#### Problem Statement:
Uncover the factors that lead to employee attrition and explore important questions such as ‘show me a breakdown of distance from home by job role and attrition’ or ‘compare average monthly income by education and attrition’. This is a fictional data set created by IBM data scientists.

In [2]:
# import required libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [3]:
# read the dataset
data = pd.read_csv('dataset/IBMEmployeeAttritioin.csv')

# display first five records
data.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [4]:
# check dimensions of the data
data.shape

(1470, 35)

In [5]:
# check datatypes present
data.dtypes

Age                          int64
Attrition                   object
BusinessTravel              object
DailyRate                    int64
Department                  object
DistanceFromHome             int64
Education                    int64
EducationField              object
EmployeeCount                int64
EmployeeNumber               int64
EnvironmentSatisfaction      int64
Gender                      object
HourlyRate                   int64
JobInvolvement               int64
JobLevel                     int64
JobRole                     object
JobSatisfaction              int64
MaritalStatus               object
MonthlyIncome                int64
MonthlyRate                  int64
NumCompaniesWorked           int64
Over18                      object
OverTime                    object
PercentSalaryHike            int64
PerformanceRating            int64
RelationshipSatisfaction     int64
StandardHours                int64
StockOptionLevel             int64
TotalWorkingYears   

In [9]:
# descriptive statistics
data.describe().transpose().round(2)

count      mean      std     min      25%      50%  \
Age                       1470.0     36.92     9.14    18.0    30.00     36.0   
DailyRate                 1470.0    802.49   403.51   102.0   465.00    802.0   
DistanceFromHome          1470.0      9.19     8.11     1.0     2.00      7.0   
Education                 1470.0      2.91     1.02     1.0     2.00      3.0   
EmployeeCount             1470.0      1.00     0.00     1.0     1.00      1.0   
EmployeeNumber            1470.0   1024.87   602.02     1.0   491.25   1020.5   
EnvironmentSatisfaction   1470.0      2.72     1.09     1.0     2.00      3.0   
HourlyRate                1470.0     65.89    20.33    30.0    48.00     66.0   
JobInvolvement            1470.0      2.73     0.71     1.0     2.00      3.0   
JobLevel                  1470.0      2.06     1.11     1.0     1.00      2.0   
JobSatisfaction           1470.0      2.73     1.10     1.0     2.00      3.0   
MonthlyIncome             1470.0   6502.93  4707.96  1009.0  2911.00   4919.0   
MonthlyRate               1470.0  14313.10  7117.79  2094.0  8047.00  14235.5   
NumCompaniesWorked        1470.0      2.69     2.50     0.0     1.00      2.0   
PercentSalaryHike         1470.0     15.21     3.66    11.0    12.00     14.0   
PerformanceRating         1470.0      3.15     0.36     3.0     3.00      3.0   
RelationshipSatisfaction  1470.0      2.71     1.08     1.0     2.00      3.0   
StandardHours             1470.0     80.00     0.00    80.0    80.00     80.0   
StockOptionLevel          1470.0      0.79     0.85     0.0     0.00      1.0   
TotalWorkingYears         1470.0     11.28     7.78     0.0     6.00     10.0   
TrainingTimesLastYear     1470.0      2.80     1.29     0.0     2.00      3.0   
WorkLifeBalance           1470.0      2.76     0.71     1.0     2.00      3.0   
YearsAtCompany            1470.0      7.01     6.13     0.0     3.00      5.0   
YearsInCurrentRole        1470.0      4.23     3.62     0.0     2.00      3.0   
YearsSinceLastPromotion   1470.0      2.19     3.22     0.0     0.00      1.0   
YearsWithCurrManager      1470.0      4.12     3.57     0.0     2.00      3.0   

                               75%      max  
Age                          43.00     60.0  
DailyRate                  1157.00   1499.0  
DistanceFromHome             14.00     29.0  
Education                     4.00      5.0  
EmployeeCount                 1.00      1.0  
EmployeeNumber             1555.75   2068.0  
EnvironmentSatisfaction       4.00      4.0  
HourlyRate                   83.75    100.0  
JobInvolvement                3.00      4.0  
JobLevel                      3.00      5.0  
JobSatisfaction               4.00      4.0  
MonthlyIncome              8379.00  19999.0  
MonthlyRate               20461.50  26999.0  
NumCompaniesWorked            4.00      9.0  
PercentSalaryHike            18.00     25.0  
PerformanceRating             3.00      4.0  
RelationshipSatisfaction      4.00      4.0  
StandardHours                80.00     80.0  
StockOptionLevel              1.00      3.0  
TotalWorkingYears            15.00     40.0  
TrainingTimesLastYear         3.00      6.0  
WorkLifeBalance               3.00      4.0  
YearsAtCompany                9.00     40.0  
YearsInCurrentRole            7.00     18.0  
YearsSinceLastPromotion       3.00     15.0  
YearsWithCurrManager          7.00     17.0

In [10]:
# concise summary of data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [11]:
# check for null/missing values
data.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

In [14]:
# check for duplicate records
data.duplicated().sum()

0

In [42]:
# function to get numerical / categorical columns
num_cols = []
cat_cols = []
def num_cat_cols(data):
    for i in data.columns:
        if data[i].dtypes != 'O':
            num_cols.append(i)
        else:
            cat_cols.append(i)

In [44]:
num_cat_cols(data)

In [45]:
num_cols

['Age',
 'DailyRate',
 'DistanceFromHome',
 'Education',
 'EmployeeCount',
 'EmployeeNumber',
 'EnvironmentSatisfaction',
 'HourlyRate',
 'JobInvolvement',
 'JobLevel',
 'JobSatisfaction',
 'MonthlyIncome',
 'MonthlyRate',
 'NumCompaniesWorked',
 'PercentSalaryHike',
 'PerformanceRating',
 'RelationshipSatisfaction',
 'StandardHours',
 'StockOptionLevel',
 'TotalWorkingYears',
 'TrainingTimesLastYear',
 'WorkLifeBalance',
 'YearsAtCompany',
 'YearsInCurrentRole',
 'YearsSinceLastPromotion',
 'YearsWithCurrManager']

In [46]:
cat_cols

['Attrition',
 'BusinessTravel',
 'Department',
 'EducationField',
 'Gender',
 'JobRole',
 'MaritalStatus',
 'Over18',
 'OverTime']

In [47]:
len(num_cols), len(cat_cols)

(26, 9)

#### Initial Data Insights:
 - The dataset has 1470 records and 35 features.
 - No missing/null/duplicate records in data.
 - There are total of 26 numerical features and 9 categorical features present.
 - Our Target Feature is "Attrition".